### DISCUSSION TITLE
Why is the DOM output so different from MathJax?

### DISCUSSION THREAD

**Discussion by ghost** [KaTeX/KaTeX#4207] [Q&A] [answered]
I am currently looking at MathJax and KaTeX and trying to see which one better suits my needs. For what I want to do, I need to have a lot of control over the individual elements in the equations.

I think KaTeX seems, at it's core, to be a bit better than MathJax (since it also is faster), but as far as I can see, with MathJax you have more control over DOM elements, since they seem to be kind of arranged in accordance with MathML structure, which would make it easier to interpret the DOM output on a hierarchical level.

So now my question is, what the intentions behind this were when developing KaTeX and is there some kind of DOM structure available in the KaTeX output (or inside of KaTeX) that I don't see yet?

**Comment by k4b7** [comment, marked as answer]
The main driving concern behind the structure of the HTML markup generated was having the appearance of the rendered math match LaTeX's output as much as possible.  Some elements of the markup generated follow TeX's box model but there are differences that were necessary to make things work with HTML.

**Comment by ghost** [comment]
Thanks for the response. Does KaTeX still have some kind of intermediate internal format (MathML etc. or custom) when processing LaTeX input though? Or does it merely 'replicate' LaTeX's typesetting behaviour in the DOM?

**Comment by k4b7** [comment]
There is a parse tree that's generated, but we convert that directly to DOM nodes and MathML nodes (we render MathML but it's hidden by default).  There was some discussion about having an intermediate representation which would be essential a platform specific description of the typeset layout of glyphs and boxes and such that could then be rendered to HTML, SVG, canvas etc.

**Comment by ghost** [comment]
So does that mean that KaTeX actually has the same typesetting steps as MathJax does? I am kind of trying to understand what it is that actually gives KaTeX this speed edge over MathJax.

**Comment by k4b7** [comment]
I believe what gives `KaTeX` its edge in terms of speed is not rendering and measuring glyphs in the browser.  The latest version of `MathJax` may not do that anymore.

**Comment by ghost** [comment]
Hey, @kevinbarabash, sorry for not responding for so long.
What do you mean by the sentence that MathJax does not measure or render glyphs in the browser? What exactly are glyphs in this context?

**Comment by k4b7** [comment]
A glyph is the visual "image" that is rendered for a particular character in a particular font.
